content/blog/2020-12-18-stp59-final.md

```diff
@@ -0,0 +1,62 @@
+
+## Implementation
+
+### Static Semantics of Parser Loops
+
+The surface syntax of P4 parsers consists of local declarations followed by states. These states are comprised of a highly restricted set of statements (variable assignments and method calls only) followed by a transition statement, indicating a jump to another state. Transitions may be conditional, branching to arbitrarily many different states. The result is a system in which arbitrary irreducible control flow arises quite easily.
+
+However, the common case is to use a natural loop in the parser to populate various data structures provided by the language. In particular, parser loops that extract from the packet on each iteration are guaranteed to terminate due to the fact that the packet has a finite length. With these considerations, I am able to provide a concrete proposal for a static analysis tool of parser loops. Accepted parsers should meet the following conditions:
+
+1) They should have reducible control flow
+2) Every distinct natural loop should include a block which both extracts from the packet and dominates all paths to the loop header
+
+My implementation runs these checks, providing the user with the options to fail when they are not satisfied or merely to emit a warning and produce unmodified code. Surprisingly, this turned out to be the most computationally heavy part of the implementation, not the actual transformation. While natural loops are easily identified about a backedge in the control flow, arbitrary loops require a full traversal of the control flow graph (CFG). I used [Tarjan's Strongly Connected Components Algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm), which boasts linear runtime, to compute strongly-connected components of an arbitrary CFG. I then check that they all satisfied the criteria for forming a natural loop: that there exists a unique header node `h` which dominates all other nodes in the strongly-connected component and which is the only node with incoming edges from outside of the loop. Having identified the natural loops, I then traverse them to ensure that every path back to the header consumes the packet. Again, this ensures termination as packets are guaranteed to be finite.
```

Comment:
Again, examples might be nice here, especially of programs that your checker would warn/error on.